controller/cmd/service-mirror/main.go

```diff
@@ -1,101 +1,160 @@
 
-	var informer cache.SharedIndexInformer
-
-	switch resource {
-	case k8s.Svc:
-		informer = sharedInformers.Core().V1().Services().Informer()
-	case k8s.Secret:
-		informer = sharedInformers.Core().V1().Secrets().Informer()
-	default:
-		return nil, fmt.Errorf("cannot instantiate local informer for %v", resource)
-
-	}
-
-	sharedInformers.Start(nil)
-
-	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
-	defer cancel()
-
-	log.Infof("waiting for local namespaced %v informer caches to sync", resource)
-	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
-		return nil, fmt.Errorf("failed to sync local namespaced %v informer caches", resource)
-	}
-	log.Infof("local namespaced %v informer  caches synced", resource)
-	return informer, nil
-}
-
-// Main executes the tap service-mirror
+// Main executes the service-mirror controller
 func Main(args []string) {
 	cmd := flag.NewFlagSet("service-mirror", flag.ExitOnError)
 
 	kubeConfigPath := cmd.String("kubeconfig", "", "path to the local kube config")
 	requeueLimit := cmd.Int("event-requeue-limit", 3, "requeue limit for events")
 	metricsAddr := cmd.String("metrics-addr", ":9999", "address to serve scrapable metrics on")
-	namespace := cmd.String("namespace", "", "address to serve scrapable metrics on")
+	namespace := cmd.String("namespace", "", "namespace containing Link and credentials Secret")
 	repairPeriod := cmd.Duration("endpoint-refresh-period", 1*time.Minute, "frequency to refresh endpoint resolution")
 
 	flags.ConfigureAndParse(cmd, args)
+	linkName := cmd.Arg(0)
 
 	stop := make(chan os.Signal, 1)
 	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
 
-	k8sAPI, err := k8s.InitializeAPI(
-		*kubeConfigPath,
-		false,
-		k8s.Svc,
-		k8s.NS,
-		k8s.Endpoint,
-	)
-
+	// We create two different kubernetes API clients for the local cluster:
+	// k8sAPI is used as a dynamic client for unstrcutured access to Link custom
+	// resources.
+	//
+	// controllerK8sAPI is used by the cluster watcher to manage
+	// mirror resources such as services, namespaces, and endpoints.
+	k8sAPI, err := k8s.NewAPI(*kubeConfigPath, "", "", []string{}, 0)
 	//TODO: Use can-i to check for required permissions
 	if err != nil {
 		log.Fatalf("Failed to initialize K8s API: %s", err)
 	}
 
-	secretsInformer, err := initLocalResourceInformer(k8sAPI.Client, *namespace, k8s.Secret)
+	controllerK8sAPI, err := controllerK8s.InitializeAPI(*kubeConfigPath, false,
+		controllerK8s.NS,
+		controllerK8s.Svc,
+		controllerK8s.Endpoint,
+	)
 	if err != nil {
-		log.Fatalf("Failed to initialize secret informer: %s", err)
+		log.Fatalf("Failed to initialize K8s API: %s", err)
 	}
-	svcInformer, err := initLocalResourceInformer(k8sAPI.Client, *namespace, k8s.Svc)
 
+	gvr := schema.GroupVersionResource{
+		Group:    "multicluster.linkerd.io",
+		Version:  "v1alpha1",
+		Resource: "links",
+	}
+	linkClient := k8sAPI.DynamicClient.Resource(gvr).Namespace(*namespace)
```

Comment:
I wonder why not add that to the api go client the same way we are doing it for TrafficSplits, etc. Why use the DynamicClient?

Reply:
We avoid using the code-generated client to have more flexible control over the deserialization and so that we can build in backwards compatibility if we later want to change the schema of the Link custom resource.